### ISSUE TITLE
[Bug]: ndcg_at_k incorrectly computes IDCG when number of retrieved docs < k, leading to inflated NDCG

### ISSUE THREAD

**Issue by saitejabandaru-in** [mlflow/mlflow#24541]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Description
In `mlflow/metrics/metric_definitions.py`, the `_ndcg_at_k_eval_fn` computes the `ndcg_at_k` score using `sklearn.metrics.ndcg_score`.

When calling `ndcg_score`, MLflow passes `k=len(retrieved[:k])` instead of the original requested `k`:
```python
score = ndcg_score(y_true, y_score, k=len(retrieved[:k]), ignore_ties=True)
```

By doing this, `scikit-learn` computes the Ideal DCG (IDCG) for `len(retrieved[:k])` documents instead of `k` documents. If a retriever returns fewer than `k` documents, the IDCG will be artificially low, which inflates the final NDCG score.

For example, if we want to compute `ndcg_at_k` at `k=5` for 5 ground-truth documents, and the retriever only returns 1 document (which happens to be relevant):
- The true NDCG@5 should be `DCG@5 / IDCG@5` = `1.0 / (1 + 1/log2(3) + ... + 1/log2(6))` ≈ `0.339`.
- However, MLflow dynamically sets `k=1` for scikit-learn. scikit-learn then computes `DCG@1 / IDCG@1` = `1.0 / 1.0` = `1.0`.

This means a model is not correctly penalized for retrieving too few relevant documents. Its NDCG@k is heavily inflated.

### Steps to Reproduce
Here is a minimal reproduction using the internal logic of `_ndcg_at_k_eval_fn`:

```python
import numpy as np
from sklearn.metrics import ndcg_score

# Suppose the user requests NDCG@5
k = 5
# There are 5 relevant documents in the ground truth
targets = ["doc1", "doc2", "doc3", "doc4", "doc5"]
# The retriever only returned 1 document
predictions = ["doc1"]

# This simulates `_prepare_row_for_ndcg` behavior
y_score = np.array([[0.999999, 0, 0, 0, 0]]) 
y_true = np.array([[1, 1, 1, 1, 1]])

# MLflow passes k=len(retrieved[:k]), which is 1
mlflow_score = ndcg_score(y_true, y_score, k=len(predictions[:k]), ignore_ties=True)
print(f"MLflow computed NDCG@{k}: {mlflow_score}")
# Output: 1.0

# Correct computation uses k=5
correct_score = ndcg_score(y_true, y_score, k=k, ignore_ties=True)
print(f"Correct NDCG@{k}: {correct_score}")
# Output: 0.3391602052736161
```

### Proposed Fix
To correctly use `sklearn.metrics.ndcg_score` for IR metrics with partial lists, `y_score` must ensure unretrieved documents are not ranked in the top `k`, while still evaluating with the original `k`.

In `_prepare_row_for_ndcg`, unretrieved documents are assigned a score of `0`. The retrieved documents are assigned scores `1 - i * eps`.
If we evaluate with `k=k`, `scikit-learn` might include the unretrieved documents (score `0`) in the top `k` if the list of retrieved documents is shorter than `k` (because it ranks 0s as ties).

A robust fix would be:
1. Assign a negative score to unretrieved documents (e.g. `-1`) so they are ranked last.
2. Ensure there are enough "dummy" non-relevant documents so that unretrieved ground-truth documents never fall into the top `k`.
3. Pass `k=k` to `ndcg_score` instead of `k=len(retrieved[:k])`.

Alternatively, given the edge cases with `scikit-learn`'s `ndcg_score` for partial rankings, it might be simpler to compute DCG and IDCG manually in `_ndcg_at_k_eval_fn` without relying on `scikit-learn`.

**Comment by adityaanikam** [comment]
Hi @saitejabandaru-in — thanks for the thorough writeup and repro, this made
the root cause very clear.

I'd like to submit a PR for this. My plan:

- In `_prepare_row_for_ndcg` (mlflow/metrics/metric_definitions.py), assign
  unretrieved ground-truth docs a score that guarantees they can't land in
  the top-k as ties with the zero-filled slots, so evaluating with the full
  original `k` doesn't let sklearn silently shrink the ranking window.
- Change `_ndcg_at_k_eval_fn` to pass `k=k` to `ndcg_score` instead of
  `k=len(retrieved[:k])`.
- Add a regression test in tests/metrics/test_metric_definitions.py using
  your repro case (5 ground-truth docs, 1 retrieved, expect NDCG@5 ≈ 0.339,
  not 1.0), plus a case where len(retrieved) >= k to confirm no regression
  there.

Could a maintainer take a look and apply the `ready` label if this seems
like the right direction? Happy to adjust the approach first if there's a
preferred fix (e.g. computing DCG/IDCG manually instead of leaning on
sklearn's ndcg_score for partial rankings).

**Comment by AkshayKarthickMS** [comment]
Hi @adityaanikam are you still planning to work on this? I don't see a PR yet. If not, I'd be happy to take it. I've reproduced the bug on current master (NDCG@5 returns 1.0 instead of ~0.339 for the example above) and have a fix ready that computes binary-relevance DCG/IDCG directly, with IDCG always taken over min(k, |relevant|) positions. All existing test_ndcg_at_k cases pass unchanged, and I've added a regression test.

@maintainers, since ndcg_at_k is deprecated as of 3.4.0, is a fix still welcome here? If so, could this get the ready label?

**Comment by adityaanikam** [comment]
Hi @AkshayKarthickMS, no, I haven't started a PR, so please go ahead. Computing DCG and IDCG directly is a better approach than what I had planned. I don't think pushing unretrieved docs down with lower scores would have worked, since with only k documents in the list they still end up in the top k. Thanks for checking first.

**Comment by AkshayKarthickMS** [comment]
Thanks @adityaanikam! I've opened #26125  with the fix (DCG/IDCG computed directly; NDCG@5 for the example goes from 1.0 to ~0.339, existing tests unchanged). Could a maintainer take a look, and apply ready if this is welcome? I'm aware ndcg_at_k is deprecated, so if fixes there aren't wanted, just let me know.
